molgenis-core-ui/src/main/java/org/molgenis/ui/metadataeditor/MetadataEditorController.java

```diff
@@ -0,0 +1,144 @@
+	{
+		super(URI);
+		this.metadataService = requireNonNull(metadataService);
+		this.entityTypeMapper = requireNonNull(entityTypeMapper);
+		this.attributeMapper = requireNonNull(attributeMapper);
+	}
+
+	@RequestMapping(method = GET)
+	public String init()
+	{
+		return "view-metadataeditor";
+	}
+
+	@RequestMapping(value = "/entityType/{id:.+}", method = GET)
+	@ResponseBody
+	public EditorEntityTypeResponse getEntityType(@PathVariable("id") String entityTypeId)
+	{
+		// metadataService.getEntityType cannot be used due to https://github.com/molgenis/molgenis/issues/5783
+		EntityType entityType = metadataService
+				.getRepository(EntityTypeMetadata.ENTITY_TYPE_META_DATA, EntityType.class).findOneById(entityTypeId);
+
+		return createEntityTypeResponse(entityType);
+	}
+
+	// FIXME currently throws an exception until entity type has an automatically generated identifier
+	@RequestMapping(value = "/create/entityType", method = GET)
+	@ResponseBody
+	public EditorEntityTypeResponse createEntityType()
+	{
+		return createEntityTypeResponse();
+	}
+
+	@RequestMapping(value = "/entityType", method = POST)
+	@ResponseStatus(OK)
+	public void upsertEntityType(@RequestBody EditorEntityType editorEntityType)
+	{
+		EntityType entityType = entityTypeMapper.toEntityType(editorEntityType);
+		metadataService.upsertEntityType(entityType);
+	}
+
+	@RequestMapping(value = "/create/attribute", method = GET)
+	@ResponseBody
+	public EditorAttributeResponse createAttribute()
+	{
+		return createAttributeResponse();
+	}
+
+	@ExceptionHandler(UnknownEntityException.class)
+	@ResponseBody
+	@ResponseStatus(HttpStatus.BAD_REQUEST)
+	public ErrorMessageResponse handleUnknownEntityException(UnknownEntityException e)
+	{
+		LOG.debug("", e);
+		return new ErrorMessageResponse(
+				Collections.singletonList(new ErrorMessageResponse.ErrorMessage(e.getMessage())));
+	}
+
+	@ExceptionHandler(RuntimeException.class)
+	@ResponseBody
+	@ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
+	public ErrorMessageResponse handleRuntimeException(RuntimeException e)
+	{
+		LOG.error("", e);
+		return new ErrorMessageResponse(
+				Collections.singletonList(new ErrorMessageResponse.ErrorMessage(e.getMessage())));
+	}
+
+	private EditorEntityTypeResponse createEntityTypeResponse()
+	{
+		EditorEntityType editorEntityType = entityTypeMapper.createEditorEntityType();
+		return createEntityTypeResponse(editorEntityType);
+	}
+
+	private EditorEntityTypeResponse createEntityTypeResponse(EntityType entityType)
+	{
+		EditorEntityType editorEntityType = entityTypeMapper.toEditorEntityType(entityType);
+		return createEntityTypeResponse(editorEntityType);
+	}
+
+	private EditorEntityTypeResponse createEntityTypeResponse(EditorEntityType editorEntityType)
+	{
+		ImmutableList<String> languageCodes = ImmutableList.copyOf(getLanguageCodes().iterator());
+		return EditorEntityTypeResponse.create(editorEntityType, languageCodes);
+	}
+
+	private EditorAttributeResponse createAttributeResponse()
```

Comment:
Similar to the EditorEntityType comment: Why take an Attribute containing all the data you need and changing it into something else?